internal/image/validate.go

```diff
@@ -127,3 +135,67 @@ func resolveAndSetImageUrl(url string, asi *application_snapshot_image.Applicati
 
 	return resolved, nil
 }
+
+func determineAttestationTime(ctx context.Context, attestations []oci.Signature) context.Context {
+	if len(attestations) == 0 {
+		log.Debug("No attestations provided to determine attestation time")
+		return ctx
+	}
+
+	pointer, err := jsonpointer.Parse("/predicate/metadata/buildFinishedOn")
+	if err != nil {
+		log.Debugf("Failed to parse the fixed JSON Pointer: %v", err)
+		panic(err)
+	}
+
+	times := make([]time.Time, 0, len(attestations))
+	for i, attestation := range attestations {
+		payload, err := cosignPolicy.AttestationToPayloadJSON(ctx, "slsaprovenance", attestation)
+		if err != nil {
+			log.Debugf("Attestation at %d is not a SLSA Provenance attestation", i)
+			continue
+		}
+
+		data := map[string]any{}
+		if err := json.Unmarshal(payload, &data); err != nil {
+			log.Debugf("Unable to unmarshall SLSA Provenance attestation at %d", i)
+			continue
+		}
+
+		maybeFinishTime, err := pointer.Eval(data)
+		if err != nil {
+			log.Debugf("Failed to evaluate JSON Pointer %s for attestation at %d", pointer, i)
+			continue
+		}
+
+		finishTime, ok := maybeFinishTime.(string)
+		if !ok {
+			log.Debugf("Unexpected buildFinishedOn value for attestation at %d: %v", i, maybeFinishTime)
+			continue
+		}
+
+		time, err := time.Parse(time.RFC3339, finishTime)
+		if err != nil {
+			log.Debugf("Unable to parse buildFinishedOn `%s` as RFC3339 time of attestation at %d", finishTime, i)
+			continue
+		}
+
+		times = append(times, time.UTC())
+	}
+
+	if len(times) == 0 {
+		return ctx
+	}
+
+	sort.Slice(times, func(i, j int) bool {
+		return times[i].After(times[j])
+	})
+
+	attestationTime := times[0]
+
+	if log.IsLevelEnabled(log.DebugLevel) {
+		log.Debugf("Determined attestation time: %s", attestationTime.Format(time.RFC3339))
+	}
+
+	return context.WithValue(ctx, policy.AttestationTime, &attestationTime)
```

Comment:
Passing parameters in the context is usually not something we want to do unless we have to. Is this really necessary? An explicit setter on the Policy could be an alternative.

Reply:
Yeah, let me refactor that